Fix sfdp_find_addr_region algorithm

boraozgen opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

`sfdp_find_addr_region()` was causing issues with SPI flashes with sector table parsed from SFDP (in particular SST26VF016B).

In particular, it was returning -1 when address 0 is passed (probably also if the address in the first region). I do not  know why the search algorithm is written to search from the higher to lower regions, but it was obvious that it would fail for the first region. Also it was harder to read due to the index manipulation.

I tested the new algorithm with SST26VF016B and it returns the correct region for the addresses I provided. Unfortunately I do not have a test setup where I can test multiple flash chips. I guess there are no automated tests for SPIF and SFDP components, maybe it is a good idea to build a CI setup for this.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [X] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 reviewed (dismissed):
Please add details to  the commit message (from pull request description)

evedon reviewed (dismissed):
Thank you for the fix.
The issue of the original code was that the last return should have been `return 0` instead of `return -1`. But I agree that it is easier to follow the logic when searching from the lower to the higher regions so let's keep your version.

I think that there are still two issues with the function though:
1- `offset` type should be `bd_addr_t` instead of `bd_size_t`. They both map to uint64_t.
2- The first check should return -1 if `offset >= sfdp_info.bptbl.device_size_bytes`

LDong-Arm commented on drivers/source/SFDP.cpp:
```diff
@@ -377,20 +377,20 @@ int sfdp_detect_erase_types_inst_and_size(uint8_t *bptbl_ptr, sfdp_hdr_info &sfd
 }
 
-int sfdp_find_addr_region(bd_size_t offset, const sfdp_hdr_info &sfdp_info)
+int sfdp_find_addr_region(bd_addr_t offset, const sfdp_hdr_info &sfdp_info)
```
Could you also update the declaration? https://github.com/ARMmbed/mbed-os/blob/798c3c5f4f7aa63210d90a43e4c1c920efcf69ed/drivers/include/drivers/internal/SFDP.h#L132

boraozgen replied:
Done

LDong-Arm reviewed (commented):
> 1- `offset` type should be `bd_addr_t` instead of `bd_size_t`. They both map to uint64_t.

Well, throughout Mbed OS storage APIs an "offset" is declared as an "address" in some places and a "size" in others.

LDong-Arm reviewed (approved):
Thanks @boraozgen, looks good to me

boraozgen committed:
Fix sfdp_find_addr_region algorithm

sfdp_find_addr_region() was causing issues with SPI
flashes with sector table parsed from SFDP (in
particular SST26VF016B).

In particular, it was returning -1 when address 0 is
passed (probably also if the address in the first
region). I do not know why the search algorithm is
written to search from the higher to lower regions,
but it was obvious that it would fail for the first
region. Also it was harder to read due to the index
manipulation.
